### PR TITLE
Add hexagonal container

### PR DESCRIPTION
**Pull request by harshavb**
Thanks to Poyo for all the help. Optimally uses 4 triangles per hexagon.

### CODE REVIEW

**Comment by bdach** [comment]
Do you have permission from @Poyo-SSB to PR this back upstream? Kind of dodgy that you're taking authorship credit on the commits, too.

For the record, original implementation was made in https://github.com/harshavb/piously/pull/13 and https://github.com/harshavb/piously/pull/21.

**Comment by harshavb** [comment]
Fair enough. Will re-open if granted permission.

**Comment by Poyo-SSB** [comment]
Permission granted. I can make adjustments and do rewrites as necessary.

**Comment by bdach** [comment]
reopening for review then, I guess.

@Poyo-SSB would you like (co-)authorship credit on the existing commits? probably one of the rare cases we could allow for a force-push.

**Comment by Poyo-SSB** [comment]
I don't mind either way. @harshavb can take care of that, probably.

File: osu.Framework/Graphics/Containers/HexagonalContainer.cs [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+using osuTK.Graphics.ES30;
+using osuTK.Graphics;
+using osuTK;
+
+namespace osu.Framework.Graphics.Containers
+{
+    public class HexagonalContainer : HexagonalContainer<Drawable> { }
+
+    public class HexagonalContainer<T> : BufferedContainer<T>
+        where T : Drawable
```
Comment (bdach) [inline osu.Framework/Graphics/Containers/HexagonalContainer.cs:21]: is this generic all that necessary? why not just stick with `HexagonalContainer`?

File: osu.Framework/Graphics/Containers/HexagonalContainer.cs [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    {
+        private static readonly float sin_pi_over_3 = (float)Math.Sin(Math.PI / 3);
```
Comment (bdach) [inline osu.Framework/Graphics/Containers/HexagonalContainer.cs:23]: fancy way of writing `0.5f`, but sure

File: osu.Framework/Graphics/Containers/HexagonalContainer.cs [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+        private static readonly float tan_pi_over_3 = (float)Math.Tan(Math.PI / 3);
+
+        public static readonly float HEXAGON_INRADIUS = sin_pi_over_3;
```
Comment (bdach) [inline osu.Framework/Graphics/Containers/HexagonalContainer.cs:26]: unused

File: osu.Framework/Graphics/Containers/HexagonalContainer.cs [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+        [BackgroundDependencyLoader]
+        private void load(ShaderManager shaders) => Shader = shaders.Load(VertexShaderDescriptor.TEXTURE_2, "TextureHexagon");
+
+        protected override DrawNode CreateDrawNode() => new HexagonalContainerDrawNode(this, sharedData);
+
+        public override bool ReceivePositionalInputAt(Vector2 screenSpacePos)
+        {
+            Vector2 norm = screenSpacePos - ScreenSpaceDrawQuad.TopLeft;
+            norm = new Vector2(norm.X / ScreenSpaceDrawQuad.Width, norm.Y / ScreenSpaceDrawQuad.Height); // apparently we can't divide Vector2s?
```
Comment (bdach) [inline osu.Framework/Graphics/Containers/HexagonalContainer.cs:43]: it is actually possible, just not via operator overload

```suggestion
            norm = Vector2.Divide(norm, ScreenSpaceDrawQuad.Size);
```

File: osu.Framework/Graphics/Containers/HexagonalContainer.cs [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+                FrameBuffer current = SharedData.CurrentEffectBuffer;
+                FrameBuffer target = SharedData.GetNextEffectBuffer();
+
+                GLWrapper.SetBlend(BlendingParameters.None);
+
+                using (BindFrameBuffer(target))
+                {
+                    float resolution = Math.Max(Source.ScreenSpaceDrawQuad.Width, Source.ScreenSpaceDrawQuad.Height); // shh
+                    hexagonShader.GetUniform<float>("g_Resolution").UpdateValue(ref resolution);
```
Comment (bdach) [inline osu.Framework/Graphics/Containers/HexagonalContainer.cs:104]: what's with this comment?

File: osu.Framework/Graphics/Shapes/Hexagon.cs [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+            {
+                Primitives.Hexagon drawingHex = toHexagon(ScreenSpaceDrawQuad);
+                DrawTriangle(Texture, drawingHex.FarUpTriangle, DrawColourInfo.Colour, null, null,
+                    new Vector2(InflationAmount.X / DrawRectangle.Width, InflationAmount.Y / DrawRectangle.Height), TextureCoords);
+                DrawTriangle(Texture, drawingHex.NearUpTriangle, DrawColourInfo.Colour, null, null,
+                    new Vector2(InflationAmount.X / DrawRectangle.Width, InflationAmount.Y / DrawRectangle.Height), TextureCoords);
+                DrawTriangle(Texture, drawingHex.NearDownTriangle, DrawColourInfo.Colour, null, null,
+                    new Vector2(InflationAmount.X / DrawRectangle.Width, InflationAmount.Y / DrawRectangle.Height), TextureCoords);
+                DrawTriangle(Texture, drawingHex.FarDownTriangle, DrawColourInfo.Colour, null, null,
+                    new Vector2(InflationAmount.X / DrawRectangle.Width, InflationAmount.Y / DrawRectangle.Height), TextureCoords);
```
Comment (bdach) [inline osu.Framework/Graphics/Shapes/Hexagon.cs:50]: can these be a method at least or something? duplicated 8 times across 2 methods, only differing with the triangle involved.

**Comment by harshavb** [comment]
Poyo should take full authorship credit, as all the code here is written by her.

**Review by bdach — CHANGES_REQUESTED** [review]

**Comment by bdach** [comment]
oh, and this PR should also include the test scene.

**Comment by Poyo-SSB** [comment]
`HexagonalContainer` is generic for the same reason that other Containers are generic—for ease of access to single-type children, yes? I don't know if there are any performance optimizations that come with that or anything.

As for the comment, I'm unsure if passing in the resolution as a uniform is the canonical method of having screenspace-aware shaders. I wasn't able to find the way that CompositeDrawable does this.

**Comment by bdach** [comment]
re: generic - i mean sure, i get that, it's just that it's generating quite a few CI inspections so i was wondering if it can be removed entirely. i guess it can stay for convenience though, the static fields can be moved to the non-generic base as protected.

not sure about the screenspace awareness. would have to poke `CompositeDrawable` myself.

**Comment by peppy** [comment]
I'd remove generics for simplicity. `Container` is an exception since it is such a commonly used class. this is not.

**Comment by bdach** [comment]
@harshavb did you happen to uncheck the "Allow edits from maintainers" checkbox on this PR? i have a batch of changes that are supposed to resolve my reviews, fix up style somewhat and reattribute previous commits, but i cannot push to this PR.

that said it might also be due to this being PRed from `master`, which is never a good idea.

**Comment by harshavb** [comment]
@bdach It's likely because it's PRed from master, since it's not unchecked.

**Comment by harshavb** [comment]
Moved to #4091

**Pull request closed without merging**